File: apps/backend/src/wise_workbench/adapters/engine/flow.py
"""FlowGraph JSON: a directly-follows graph with metrics, stage groups and constraint overlays.

The payload follows the flow library's contract (``@wise/flow`` ``docs/API.md``):
node metrics ``cases``, ``events``, ``share``, ``violationShare``; follows-edge
metrics ``count``, ``cases``, ``share``, ``medianLagHours``; overlay payloads
with ``value`` (a share in [0, 1]), ``label``, ``text``, ``coverage``, ``glyph``,
``threshold``, ``unit``. The workbench's own names (``caseShare``,
``medianHours``, ``shareViolated``) stay next to them.
"""

from __future__ import annotations

import re
from typing import TYPE_CHECKING, Any

import pandas as pd
import wise

if TYPE_CHECKING:  # pragma: no cover
    from wise_workbench.adapters.knowledge import StageModel

START, END = "__start__", "__end__"


def _node_id(label: str) -> str:
    slug = re.sub(r"[^A-Za-z0-9]+", "_", label).strip("_").lower() or "node"
    return f"a_{slug}"


def _pct(value: float) -> str:
    p = value * 100
    return f"{p:.0f} %" if p >= 10 else f"{p:.1f} %"


def _payload(nc: wise.NormConstraint, share: dict[str, float], n_cases: int) -> dict[str, Any]:
    c = nc.constraint
    evaluated: int = int(share.get("casesEvaluated") or 0)
    value = share.get("shareViolated")
    description = str(nc.description or nc.id)
    if value is not None and evaluated > 0:
        text = f"{description.rstrip('.')} — missed in {_pct(float(value))} of {evaluated:,} evaluated cases"
    else:
        text = f"{description.rstrip('.')} — not evaluated in this scene"
    payload: dict[str, Any] = {
        "constraintId": nc.id,
        "constraintType": c.type,
        "type": c.type,
        "layer": nc.layer,
        "label": nc.id,
        "description": description,
        "value": value,
        "coverage": (evaluated / n_cases) if evaluated > 0 else None,
        "text": text,
        **share,
    }
    if isinstance(c, wise.Presence):
        payload.update(glyph=f"≥{c.m}", threshold=float(c.m))
    elif isinstance(c, wise.Singularity):
        payload.update(glyph=f"≤{c.k}", threshold=float(c.k), width=float(c.K))
    elif isinstance(c, wise.Exclusion):
        payload.update(glyph="∅", threshold=0.0)
    elif isinstance(c, wise.Lag):
        payload.update(
            glyph="⇒",
            threshold=float(c.delta) if c.delta is not None else None,
            width=float(c.width),
            unit=str(c.unit),
        )
    elif isinstance(c, wise.Precedence):
        payload.update(glyph="⇒", threshold=float(c.k), width=float(c.K))
    elif isinstance(c, wise.Balance):
        payload.update(glyph="=", threshold=float(c.tau), width=float(c.width))
    elif isinstance(c, wise.Metric):
        payload.update(
            glyph="#", threshold=float(c.threshold), width=float(c.width), unit=str(c.attribute), direction=c.direction
        )
    return payload


def build_flow_graph(
    dfg: dict[str, Any],
    norm: wise.Norm | None,
    violation_shares: dict[str, dict[str, float]],
    *,
    abstraction: float,
    meta: dict[str, Any],
    stages: StageModel | None = None,
) -> dict[str, Any]:
    """``dfg`` comes from the storage adapter; ``violation_shares`` maps constraint id → {shareViolated, meanViolation, casesEvaluated}."""
    n_cases = max(int(dfg.get("cases", 0)), 1)
    node_rows = [n for n in dfg["nodes"] if n["cases"] / n_cases >= abstraction]
    kept = {n["label"] for n in node_rows}
    ids = {label: _node_id(label) for label in kept}
    # keep ids unique when two labels slug to the same id
    seen: dict[str, int] = {}
    for label in sorted(kept):
        base = ids[label]
        if base in seen:
            seen[base] += 1
            ids[label] = f"{base}_{seen[base]}"
        else:
            seen[base] = 0
    constraints = list(norm.constraints) if norm is not None else []
    # violation share per activity: the largest share among the expectations that mention it
    node_violation: dict[str, float] = {}
    for nc in constraints:
        node_share = violation_shares.get(nc.id, {}).get("shareViolated")
        if node_share is None:
            continue
        for a in nc.constraint.activities():
            node_violation[a] = max(node_violation.get(a, 0.0), float(node_share))
    nodes: list[dict[str, Any]] = [
        {
            "id": START,
            "kind": "event",
            "label": "start",
            "group": None,
            "metrics": {"cases": float(n_cases), "share": 1.0},
            "tags": ["start"],
        },
        {
            "id": END,
            "kind": "event",
            "label": "end",
            "group": None,
            "metrics": {"cases": float(n_cases), "share": 1.0},
            "tags": ["end"],
        },
    ]
    group_members: dict[str, int] = {}
    for n in sorted(node_rows, key=lambda r: -r["events"]):
        stage = stages.stage_of(n["label"]) if stages is not None else None
        if stage:
            group_members[stage] = group_members.get(stage, 0) + 1
        nodes.append(
            {
                "id": ids[n["label"]],
                "kind": "activity",
                "label": n["label"],
                "group": stage,
                "metrics": {
                    "events": float(n["events"]),
                    "cases": float(n["cases"]),
                    "share": n["cases"] / n_cases,
                    "caseShare": n["cases"] / n_cases,
                    "eventsPerCase": n["events"] / max(n["cases"], 1),
                    "violationShare": node_violation.get(n["label"], 0.0),
                },
                "tags": [],
            }
        )
    edges: list[dict[str, Any]] = []
    for e in dfg["edges"]:
        if e["source"] not in kept or e["target"] not in kept or e["cases"] / n_cases < abstraction:
            continue
        metrics = {
            "count": float(e["count"]),
            "cases": float(e["cases"]),
            "share": e["cases"] / n_cases,
            "caseShare": e["cases"] / n_cases,
        }
        if e.get("median_hours") is not None:
            metrics["medianLagHours"] = float(e["median_hours"])
            metrics["medianHours"] = float(e["median_hours"])
        edges.append(
            {
                "id": f"f_{ids[e['source']]}__{ids[e['target']]}",
                "kind": "follows",
                "source": ids[e["source"]],
                "target": ids[e["target"]],
                "metrics": metrics,
                "tags": ["selfLoop"] if e["source"] == e["target"] else [],
            }
        )
    for label, n in dfg.get("starts", {}).items():
        if label in kept and n / n_cases >= abstraction:
            edges.append(
                {
                    "id": f"f_start__{ids[label]}",
                    "kind": "follows",
                    "source": START,
                    "target": ids[label],
                    "metrics": {"count": float(n), "cases": float(n), "share": n / n_cases, "caseShare": n / n_cases},
                    "tags": ["start"],
                }
            )
    for label, n in dfg.get("ends", {}).items():
        if label in kept and n / n_cases >= abstraction:
            edges.append(
                {
                    "id": f"f_{ids[label]}__end",
                    "kind": "follows",
                    "source": ids[label],
                    "target": END,
                    "metrics": {"count": float(n), "cases": float(n), "share": n / n_cases, "caseShare": n / n_cases},
                    "tags": ["end"],
                }
            )
    overlays: list[dict[str, Any]] = []
    without_activity: list[str] = []
    constraints_meta: list[dict[str, Any]] = []
    for nc in constraints:
        c = nc.constraint
        share = violation_shares.get(nc.id, {})
        payload = _payload(nc, share, n_cases)
        acts = [a for a in c.activities() if a in kept]
        descr: dict[str, Any] = {
            "id": nc.id,
            "type": c.type,
            "label": nc.id,
            "layer": nc.layer,
            "description": nc.description,
            "activities": [ids[a] for a in acts],
            "params": {k: v for k, v in payload.items() if k in ("threshold", "width", "unit", "direction")},
        }
        if isinstance(c, wise.Lag | wise.Precedence):
            sources = [a for a in c.a if a in kept]
            targets = [a for a in c.b if a in kept]
            descr["a"] = [ids[a] for a in sources]
            descr["b"] = [ids[b] for b in targets]
            for a in sources:
                for b in targets:
                    eid = f"c_{nc.id}__{ids[a]}__{ids[b]}"
                    arc_payload = {**payload, "source": ids[a], "target": ids[b]}
                    edges.append(
                        {
                            "id": eid,
                            "kind": "constraint",
                            "source": ids[a],
                            "target": ids[b],
                            "metrics": {k: float(v) for k, v in share.items()},
                            "tags": [c.type],
                            "payload": arc_payload,
                        }
                    )
                    overlays.append({"kind": "arc", "target": eid, "payload": arc_payload})
            if not sources or not targets:
                without_activity.append(nc.id)
        elif isinstance(c, wise.Exclusion):
            for a in acts:
                overlays.append({"kind": "hatch", "target": ids[a], "payload": payload})
        elif isinstance(c, wise.Presence | wise.Singularity):
            for a in acts:
                overlays.append({"kind": "badge", "target": ids[a], "payload": payload})
        elif isinstance(c, wise.Balance):
            for a in acts:
                overlays.append({"kind": "chip", "target": ids[a], "payload": payload})
        else:
            without_activity.append(nc.id)
        if share.get("shareViolated"):
            for a in acts:
                overlays.append(
                    {
                        "kind": "tint",
                        "target": ids[a],
                        "payload": {
                            "constraintId": nc.id,
                            "label": nc.id,
                            "value": share["shareViolated"],
                            "shareViolated": share["shareViolated"],
                        },
                    }
                )
        constraints_meta.append(
            {
                "description": descr,
                "stats": {
                    "cases": n_cases,
                    "evaluated": share.get("casesEvaluated"),
                    "violationShare": share.get("shareViolated"),
                    "coverage": payload.get("coverage"),
                    "meanViolation": share.get("meanViolation"),
                },
            }
        )
    groups: list[dict[str, Any]] = []
    if stages is not None:
        for stage_def in stages.stages:
            if group_members.get(str(stage_def["id"])):
                groups.append(
                    {"id": str(stage_def["id"]), "kind": "stage", "label": str(stage_def["label"]), "parent": None}
                )
    out_meta = {
        **meta,
        "abstraction": abstraction,
        "cases": dfg.get("cases", 0),
        "events": dfg.get("events", 0),
        "nodesTotal": len(dfg["nodes"]),
        "edgesTotal": len(dfg["edges"]),
        "constraintsWithoutNodes": without_activity,
        "stages": [dict(s) for s in stages.stages] if stages is not None else [],
        "stagedActivities": sum(group_members.values()),
        "constraints": constraints_meta,
    }
    return {"nodes": nodes, "edges": edges, "groups": groups, "overlays": overlays, "meta": out_meta}


def violation_shares(violations: pd.DataFrame) -> dict[str, dict[str, float]]:
    out: dict[str, dict[str, float]] = {}
    for cid in violations.columns:
        v = violations[cid]
        ev = v.notna()
        if not ev.any():
            out[str(cid)] = {}
            continue
        out[str(cid)] = {
            "shareViolated": float((v[ev] > 0).mean()),
            "meanViolation": float(v[ev].mean()),
            "casesEvaluated": int(ev.sum()),
        }
    return out

File: apps/frontend/src/mocks/fixtures/flow.ts
import type { FlowGraph } from "@wise/api-schema";
import { bpic19Norm } from "./norm";
import { rng, round } from "./seed";

interface Activity {
  id: string;
  label: string;
  stage: string;
  cases: number;
  events: number;
}

const STAGES = [
  { id: "ordering", label: "Ordering", order: 1 },
  { id: "receiving", label: "Receiving", order: 2 },
  { id: "invoicing", label: "Invoicing", order: 3 },
  { id: "payment", label: "Payment", order: 4 },
];

const ACTIVITIES: Activity[] = [
  { id: "a_create_purchase_order_item", label: "Create Purchase Order Item", stage: "ordering", cases: 251734, events: 251734 },
  { id: "a_change_price", label: "Change Price", stage: "ordering", cases: 11240, events: 12988 },
  { id: "a_change_quantity", label: "Change Quantity", stage: "ordering", cases: 8402, events: 9021 },
  { id: "a_record_goods_receipt", label: "Record Goods Receipt", stage: "receiving", cases: 234479, events: 314097 },
  { id: "a_record_service_entry_sheet", label: "Record Service Entry Sheet", stage: "receiving", cases: 22314, events: 45872 },
  { id: "a_vendor_creates_invoice", label: "Vendor creates invoice", stage: "invoicing", cases: 214920, events: 224117 },
  { id: "a_record_invoice_receipt", label: "Record Invoice Receipt", stage: "invoicing", cases: 214412, events: 228760 },
  { id: "a_cancel_invoice_receipt", label: "Cancel Invoice Receipt", stage: "invoicing", cases: 5920, events: 6312 },
  { id: "a_remove_payment_block", label: "Remove Payment Block", stage: "payment", cases: 58811, events: 61230 },
  { id: "a_clear_invoice", label: "Clear Invoice", stage: "payment", cases: 208893, events: 217480 },
];

const FOLLOWS: [string, string, number, number][] = [
  ["__start__", "a_create_purchase_order_item", 251734, 0],
  ["a_create_purchase_order_item", "a_record_goods_receipt", 180211, 12 * 24],
  ["a_create_purchase_order_item", "a_vendor_creates_invoice", 41830, 9 * 24],
  ["a_create_purchase_order_item", "a_change_price", 9800, 26],
  ["a_create_purchase_order_item", "a_record_service_entry_sheet", 18102, 15 * 24],
  ["a_change_price", "a_record_goods_receipt", 8102, 6 * 24],
  ["a_change_quantity", "a_record_goods_receipt", 6011, 5 * 24],
  ["a_create_purchase_order_item", "a_change_quantity", 7204, 30],
  ["a_record_goods_receipt", "a_record_goods_receipt", 61230, 4 * 24],
  ["a_record_goods_receipt", "a_vendor_creates_invoice", 158004, 6 * 24],
  ["a_record_service_entry_sheet", "a_vendor_creates_invoice", 17820, 7 * 24],
  ["a_vendor_creates_invoice", "a_record_invoice_receipt", 209911, 2 * 24],
  ["a_record_invoice_receipt", "a_cancel_invoice_receipt", 5920, 3 * 24],
  ["a_cancel_invoice_receipt", "a_record_invoice_receipt", 5100, 4 * 24],
  ["a_record_invoice_receipt", "a_remove_payment_block", 58811, 11 * 24],
  ["a_record_invoice_receipt", "a_clear_invoice", 132743, 36 * 24],
  ["a_remove_payment_block", "a_clear_invoice", 56102, 9 * 24],
  ["a_clear_invoice", "__end__", 208893, 0],
  ["a_record_invoice_receipt", "__end__", 14020, 0],
  ["a_record_goods_receipt", "__end__", 28821, 0],
];

const ID_BY_LABEL = Object.fromEntries(ACTIVITIES.map((a) => [a.label, a.id]));

/** A FlowGraph in the backend's shape (library metric names, stage groups, constraint overlays). */
export interface BuildFlowOptions {
  sliceKey?: string;
  slicing?: string;
  abstraction?: number;
  /** Share of the whole log the scene covers (a flow type, a filtered log); derived from the slice key when absent. */
  scale?: number;
  /** Activity id whose incoming and outgoing paths are returned in `paths` (CONTRACT_CYCLE2.md, flow additions). */
  focus?: string;
}

export function buildFlow(opts: BuildFlowOptions = {}): FlowGraph {
  const r = rng(`flow:${opts.slicing ?? ""}:${opts.sliceKey ?? "all"}`);
  const scale = opts.scale ?? (opts.sliceKey ? r.range(0.002, 0.05) : 1);
  const shift = opts.sliceKey ? r.range(1.1, 1.8) : 1;
  const abstraction = opts.abstraction ?? 0.05;
  const totalCases = Math.round(251734 * scale);
  const shares: Record<string, { shareViolated: number; meanViolation: number; casesEvaluated: number }> = {};
  for (const c of bpic19Norm.constraints) {
    const base = rng(`vio:${c.id}`).range(0.02, 0.6);
    const share = Math.min(0.99, round(base * shift, 4));
    shares[c.id] = { shareViolated: share, meanViolation: round(share * 0.8, 4), casesEvaluated: Math.round(totalCases * rng(`ev:${c.id}`).range(0.4, 1)) };
  }
  const violationOf: Record<string, number> = {};
  for (const c of bpic19Norm.constraints) {
    const acts = [...(c.params.activity as string[] | undefined) ?? [], ...((c.params.a as string[] | undefined) ?? []), ...((c.params.b as string[] | undefined) ?? [])];
    for (const a of acts) violationOf[a] = Math.max(violationOf[a] ?? 0, shares[c.id]!.shareViolated);
  }
  const kept = ACTIVITIES.filter((a) => a.cases / 251734 >= abstraction);
  const keptIds = new Set(kept.map((a) => a.id));
  const nodes: FlowGraph["nodes"] = [
    { id: "__start__", kind: "event", label: "start", metrics: { cases: totalCases, share: 1 }, tags: ["start"] },
    { id: "__end__", kind: "event", label: "end", metrics: { cases: totalCases, share: 1 }, tags: ["end"] },
    ...kept.map((a) => ({
      id: a.id,
      kind: "activity" as const,
      label: a.label,
      group: a.stage,
      metrics: {
        events: Math.round(a.events * scale),
        cases: Math.round(a.cases * scale),
        share: round(a.cases / 251734, 4),
        caseShare: round(a.cases / 251734, 4),
        eventsPerCase: round(a.events / a.cases, 3),
        violationShare: violationOf[a.label] ?? 0,
      },
      tags: [],
    })),
  ];
  const edges: FlowGraph["edges"] = FOLLOWS.filter(([s, t, n]) => (s === "__start__" || keptIds.has(s)) && (t === "__end__" || keptIds.has(t)) && n / 251734 >= abstraction).map(([s, t, n, hours]) => ({
    id: `f_${s.replace("__", "")}__${t.replace("__", "")}`,
    kind: "follows" as const,
    source: s,
    target: t,
    metrics: { count: Math.round(n * scale * 1.05), cases: Math.round(n * scale), share: round(n / 251734, 4), caseShare: round(n / 251734, 4), ...(hours ? { medianLagHours: hours, medianHours: hours } : {}) },
    tags: s === t ? ["selfLoop"] : s === "__start__" ? ["start"] : t === "__end__" ? ["end"] : [],
  }));
  const overlays: FlowGraph["overlays"] = [];
  const constraintsMeta: { description: Record<string, unknown>; stats: Record<string, unknown> }[] = [];
  for (const c of bpic19Norm.constraints) {
    const share = shares[c.id]!;
    const payload: Record<string, unknown> = {
      constraintId: c.id,
      constraintType: c.type,
      type: c.type,
      layer: c.layer,
      label: c.id,
      description: c.description,
      value: share.shareViolated,
      coverage: round(share.casesEvaluated / totalCases, 4),
      text: `${(c.description ?? c.id).replace(/\.$/, "")} — missed in ${Math.round(share.shareViolated * 100)} % of ${share.casesEvaluated.toLocaleString("en")} evaluated cases`,
      ...share,
    };
    const ids = (v: unknown) => ((v as string[] | undefined) ?? []).map((l) => ID_BY_LABEL[l]).filter((id): id is string => !!id && keptIds.has(id));
    const activities = ids(c.params.activity);
    const description: Record<string, unknown> = { id: c.id, type: c.type, label: c.id, layer: c.layer, description: c.description, activities, params: {} };
    if (c.type === "lag" || c.type === "precedence") {
      const a = ids(c.params.a);
      const b = ids(c.params.b);
      description.a = a;
      description.b = b;
      if (c.type === "lag") Object.assign(payload, { glyph: "⇒", threshold: c.params.delta, width: c.params.width, unit: c.params.unit });
      else Object.assign(payload, { glyph: "⇒", threshold: c.params.k, width: c.params.K });
      for (const s of a)
        for (const t of b) {
          const eid = `c_${c.id}__${s}__${t}`;
          edges.push({ id: eid, kind: "constraint", source: s, target: t, metrics: { ...share }, tags: [c.type], payload: { ...payload, source: s, target: t } });
          overlays.push({ kind: "arc", target: eid, payload: { ...payload, source: s, target: t } });
        }
    } else if (c.type === "exclusion") {
      Object.assign(payload, { glyph: "∅", threshold: 0 });
      for (const a of activities) overlays.push({ kind: "hatch", target: a, payload });
    } else if (c.type === "presence" || c.type === "singularity") {
      Object.assign(payload, c.type === "presence" ? { glyph: `≥${c.params.m ?? 1}`, threshold: c.params.m ?? 1 } : { glyph: `≤${c.params.k}`, threshold: c.params.k, width: c.params.K });
      for (const a of activities) overlays.push({ kind: "badge", target: a, payload });
    }
    if (share.shareViolated) for (const a of activities) overlays.push({ kind: "tint", target: a, payload: { constraintId: c.id, label: c.id, value: share.shareViolated, shareViolated: share.shareViolated } });
    constraintsMeta.push({ description, stats: { cases: totalCases, evaluated: share.casesEvaluated, violationShare: share.shareViolated, coverage: payload.coverage, meanViolation: share.meanViolation } });
  }
  const groups: FlowGraph["groups"] = STAGES.filter((s) => kept.some((a) => a.stage === s.id)).map((s) => ({ id: s.id, kind: "stage" as const, label: s.label }));
  const paths = opts.focus
    ? {
        incoming: FOLLOWS.filter(([, t]) => t === opts.focus && keptIds.has(t)).map(([s, , n, hours]) => ({ from: s, count: Math.round(n * scale * 1.05), cases: Math.round(n * scale), median_lag: hours ? round(hours / 24, 2) : null, violation_share: round(violationOf[ACTIVITIES.find((a) => a.id === s)?.label ?? ""] ?? 0, 4) })),
        outgoing: FOLLOWS.filter(([s]) => s === opts.focus && keptIds.has(s)).map(([, t, n, hours]) => ({ to: t, count: Math.round(n * scale * 1.05), cases: Math.round(n * scale), median_lag: hours ? round(hours / 24, 2) : null, violation_share: round(violationOf[ACTIVITIES.find((a) => a.id === t)?.label ?? ""] ?? 0, 4) })),
      }
    : undefined;
  return {
    nodes,
    edges,
    groups,
    overlays,
    ...(opts.focus ? { focus: opts.focus, paths } : {}),
    meta: {
      runId: "run_41",
      slicing: opts.slicing ? opts.slicing.split("+") : null,
      sliceKey: opts.sliceKey ? (JSON.parse(opts.sliceKey) as unknown) : null,
      process: "p2p",
      abstraction,
      cases: totalCases,
      events: Math.round(1595923 * scale),
      nodesTotal: 42,
      edgesTotal: 498,
      constraintsWithoutNodes: ["c_l6_high_exposure_memo", "c_l6_high_exposure_price_change", "c_l7_manual_touches", "c_l7_distinct_human_resources", "c_l7_total_events", "c_l7_manual_share", "c_l6_networth_volatility"],
      stages: STAGES,
      stagedActivities: kept.length,
      constraints: constraintsMeta,
    },
  };
}
